### ISSUE TITLE
Apparent Memory leak in Neural-Style example

### ISSUE THREAD

**Issue by NOTtheMessiah** [dmlc/MXNet.jl#56]
I ran the [Julia version of the Neural-Style example](https://github.com/NOTtheMessiah/MXNet.jl/blob/master/examples/neural-style/run.jl) with a check on GPU memory usage, and it seems to grow linearly from time to time, behavior not seen [in the Python version](https://github.com/dmlc/mxnet/tree/master/example/neural-style) despite being nearly identitical syntactically. Perhaps this is a difference between Julia and Python semantics? Or maybe adding type signatures to some arrays/composite types could help.

Note: to run the Julia fork, you need to add the Factor subtype of AbstractLearningRateScheduler to optimizer.jl in the package directory.

Attached is a plot of Epoch vs GPU Memory usage in megabytes.
![memusage](https://cloud.githubusercontent.com/assets/2946283/12240067/b082a1fa-b859-11e5-98a7-1ef949a11182.jpg)

**Comment by NOTtheMessiah** [comment]
Actually, replaced `a::NDArray /= b` with `mx.div_from!(a,b)` and `a::NDArray *= b` with `mx.mul_to!(a,b)` and I have much better performance, though some operations still seem to be using more memory than the Python equivalent.
![memusage2](https://cloud.githubusercontent.com/assets/2946283/12243120/ce9fd9e4-b86a-11e5-99e2-82a2069a67b5.png)
Didn't realize that operations such as `/=` couldn't be overloaded.

**Comment by vchuravy** [comment]
See the discussion in https://github.com/JuliaLang/julia/issues/249 and https://github.com/JuliaLang/julia/pull/13666

We should probably mention that and maybe we can use a macro to turn them into inplace operations `@nd_as_jl` might be a good place of doing that.

**Comment by NOTtheMessiah** [comment]
Thanks, I was able to squash the other major leak, although it might be worth noting that the macro didn't like being fed the contents of a composite type, but that was easy enough to work around.

**Comment by vchuravy** [comment]
It might be helpful for others to have a FAQ with things like these. Would
you mind writing up what you did?

**Comment by pluskid** [comment]
Yes, there is a potential memory "leak", as the underlying memory is managed by the GC. @NOTtheMessiah Could you test whether calling `gc()` every several iterations would help? -- it will slow down the computation, though.

The difference between Python and Julia is that Python GC is using ref-counting, which is able to re-claim the memory right after it is not being referenced somewhere. But Julia needs to wait a global GC call. I'm not familiar with the underlying GC implementation, but it might be automatically triggered when free memory is low. But the memory allocated by libmxnet (and especially GPU memory) is not visible to Julia GC, so it has no idea of when should GC be called. 

Unfortunately, as far as I know there is currently no immediately better way to do native resource management in Julia.

**Comment by NOTtheMessiah** [comment]
@pluskid what I did was write a convenience function that calls `nvidia-smi` from the shell and parses the response to get a broad overview to make the plots.

@vchuravy I wrote some notes, feel free to remix

# Porting MXNet's Neural-Style Python code to MXNet.jl

These notes intend to document what I've had to do to get a working implementation of [A Neural Algorithm of Artistic Style](http://arxiv.org/abs/1508.06576) from within Julia. These may or may not be applicable for other examples ported from Python to Julia, but are here for reference.

### MXNet.jl

* [Official documentation](http://mxnetjl.readthedocs.org/) might be a good place to start.

### argparse to ArgParse.jl

* [ArgParse.jl's documention](http://argparsejl.readthedocs.org/)

This library is used to turn it into a command-line program.

### Named tuples to Composite types

* [Julia documentation on (composite) types](https://julia.readthedocs.org/en/latest/manual/types/#composite-types)

This translation is pretty easy, and Julia's built-in composite types are very straightforward to work with, compared to Python's named tuples which require `from collections import namedtuple`. For the `ConvExecutor`, it means going from 
```python
Executor = namedtuple('Executor', ['executor', 'data', 'data_grad'])
```
to
```julia
type SGExecutor
    executor :: mx.Executor
    data :: mx.NDArray
    data_grad :: mx.NDArray
end
```
I also renamed it just to avoid ambiguity. The type signatures are optional, but I included them for type safety. 

### Row-major order to Column-major order arrays

If interested, read the wikipedia article [here](https://en.wikipedia.org/wiki/Row-major_order), but summarized, Julia tends to think of a matrix as a array of column-vectors, while Python natively stores it as a list of row lists. The major difference when it comes to dealing with multi-dimensional arrays is that the ordering of shape is inverted.

Example:
```python
out.infer_shape(data=(1, 3, input_size[0], input_size[1]))
```
would become the following in Julia

```julia
mx.infer_shape(out, data=(input_size[1], input_size[2], 3, 1))
```

### Memory leaks

As of Julia v0.4, some methods cannot be directly overwritten such as assignment operators (`_ ⊗= _` for some binary operation ⊗ in `[-,+,/,*]`), so for lines that take the form `a[:] ⊗= ...`, you can take one of several approaches to avoid unnecessary use of graphics memory, all of which are described in the [NDArray source code](https://github.com/dmlc/MXNet.jl/blob/master/src/ndarray.jl).

1. `b=copy(a::NDArray)` to a julia array then perform your operations and then `copy!(a,b)` back.
2. some combination of `mx.mult_to!`, `mx.div_from!`, etc. for directly modifying the NDArray
3. Use the macro `mx.@nd_as_jl` to work as if you were just using native arrays

This takes arguments `ro` for NDArrays that are read and `rw` for those that are written to. This macro copies everything to native Julia arrays and then writes them back into the NDArrays.

### Implement Factor-based Learning Rate

This required modifying Mocha.jl's *src/optimizer.jl* as to include an unimplemented subtype of `AbstractLearningRateScheduler` so it can cooperate with the Stochastic Gradient Descent. There may be a better way to do this. Also not entirely comfortable with the robustness of my implementation as of the time of writing.
